src/core/fair_queue.cc

```diff
@@ -318,37 +319,100 @@ fair_queue::clock_type::time_point fair_queue::next_pending_aio() const noexcept
          * which's sub-optimal. The expectation is that we think disk
          * works faster, than it really does.
          */
-        auto over = _group.capacity_deficiency(_pending->head);
+        auto over = _group.capacity_deficiency(_pending.head);
         auto ticks = _group.capacity_duration(over);
         return std::chrono::steady_clock::now() + std::chrono::duration_cast<std::chrono::microseconds>(ticks);
     }
 
     return std::chrono::steady_clock::time_point::max();
 }
 
+// This function is called by the shard on every poll.
+// It picks up tokens granted by the group, spends available tokens on IO dispatches,
+// and makes a reservation for more tokens, if needed.
+//
+// Reservations are done in batches of size `_group.per_tick_grab_threshold()`.
+// During contention, in an average moment in time each contending shard can be expected to
+// be holding a reservation of such size after the current head of the token bucket.
+//
+// A shard which is currently calling `dispatch_requests()` can expect a latency
+// of at most `nr_contenders * _group.per_tick_grab_threshold()` before its next reservation is fulfilled.
+// If a shard calls `dispatch_requests()` at least once per X total tokens, it should receive bandwidth
+// of at least `_group.per_tick_grab_threshold() / (X + nr_contenders * _group.per_tick_grab_threshold())`.
+// If shards have low CPU utilization (and so X is negligible), this should result in near-perfect fairness.
+// If shards have high CPU utilization (and so they poll rarely and X is a sizeable fraction of the bucket),
+// it still gives reasonable fairness: with task quota of 500us and IO latency goal of 750 us,
+// a CPU-starved shard should still get at least 60% of fair throughput.
 void fair_queue::dispatch_requests(std::function<void(fair_queue_entry&)> cb) {
-    capacity_t dispatched = 0;
-    boost::container::small_vector<priority_class_ptr, 2> preempt;
+    _group.maybe_replenish_capacity(_group_replenish);
```

Comment:
Is it needed here? The next call to reap_pending_capacity() calls it when it needs

Reply:
Oh, that's an oops. Originally I put the `maybe_replenish_capacity` in `reap_pending_capacity`, then decided it's cleaner to call it just once per `dispatch_requests`, so I moved it outside of `reap_pending_capacity`. There is even a comment about that:
```
    // Assumes that `_group.maybe_replenish_capacity()` was called recently.
    capacity_t reap_pending_capacity(bool& our_turn_has_come) noexcept;
```

But apparently I copy-pasted the old variant of `reap_pending_capacity` into the PR at some point, so the `maybe_replenish_capacity` call ended up being in both places.

Reply:
Oh, that's an oops. Originally I put the `maybe_replenish_capacity` in `reap_pending_capacity`, then decided it's cleaner to call it just once per `dispatch_requests`, so I moved it outside of `reap_pending_capacity`. There is even a comment about that:
```
    // Assumes that `_group.maybe_replenish_capacity()` was called recently.
    capacity_t reap_pending_capacity(bool& our_turn_has_come) noexcept;
```

But apparently I copy-pasted the old variant of `reap_pending_capacity` into the PR at some point, so the `maybe_replenish_capacity` call ended up being in both places.